File: src/main/java/enginePackage/Thing.java
package enginePackage;

import wrightglPackage.WrightGL;
import mathPackage.Mat4;
import mainPackage.Global;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;


public class Thing {
    public WrightGL.Program shaderProgram;
    public List<WrightGL.VB> vertexBuffers = new ArrayList<>();
    public Mat4 modelMatrix;
    Consumer<Thing> callback = null;


    public Thing(final WrightGL.Program _program, WrightGL.VB _vertexBuffer){
        shaderProgram = _program;
        vertexBuffers.add(_vertexBuffer);
        modelMatrix = Mat4.getIdentityMatrix();
    }
    public void setCallback(Consumer<Thing> _modelMatrixCallback){
        callback = _modelMatrixCallback;
    }
    public void render(Consumer<Thing> innerCallback){
        render(null,  innerCallback);
    }
    public void render(WrightGL.Program _program){
        render(_program,  null);
    }
    public void render(){
        render(null,  null);
    }
    public void render(WrightGL.Program _program, Consumer<Thing> innerCallback){
        // Shader
        final WrightGL.Program currProgram = _program != null ? _program : shaderProgram;
        currProgram.use();

        if (innerCallback != null)
            innerCallback.accept(this);

        // Update model matrix
        if (callback != null)
            callback.accept(this);

        // Uniforms
        Global.engine.wgl.setUniform("M", modelMatrix.vals, WrightGL.UniformType.Matrix);

        // Render
        for (WrightGL.VB buffer: vertexBuffers){
            buffer.render();
        }
    }
}

File: src/main/java/_0_1/enginePackage/Thing.kt
package _0_1.enginePackage

import _0_1.mathPackage.Mat4.Companion.identityMatrix
//import _0_1.wrightglPackage.WrightGL.Program.use
//import _0_1.wrightglPackage.WrightGL.setUniform
//import _0_1.wrightglPackage.WrightGL.VB.render
import _0_1.wrightglPackage.Program
import _0_1.wrightglPackage.VB
import _0_1.mathPackage.Mat4
import _0_1.mainPackage.Global
import kotlin.jvm.JvmOverloads
import _0_1.wrightglPackage.WrightGL
import java.util.ArrayList
import java.util.function.Consumer

class Thing(var shaderProgram: Program, _vertexBuffer: VB) {
    var vertexBuffers: MutableList<VB> = ArrayList()
    var modelMatrix: Mat4
    var callback: ((Thing) -> Unit)? = null
//        set(_modelMatrixCallback: ((Thing) -> Void)?){
//            field = _modelMatrixCallback
//        }
//        get() = field
//    fun setCallback(_modelMatrixCallback: Consumer<Thing>?) {
//        callback = _modelMatrixCallback
//    }


    @JvmOverloads
    fun render(_program: Program? = null, innerCallback: Consumer<Thing?>? = null) {
        // Shader
        val currProgram = _program ?: shaderProgram
        currProgram.use()
        innerCallback?.accept(this)

        // Update model matrix
        if (callback != null) callback!!(this)

        // Uniforms
        Global.engine!!.wgl.setUniform("M", modelMatrix.vals, WrightGL.UniformType.Matrix)

        // Render
        for (buffer in vertexBuffers) {
            buffer.render()
        }
    }

    init {
        vertexBuffers.add(_vertexBuffer)
        modelMatrix = identityMatrix
    }
}